File: backends/nest/src/adapter/memory.ts
// In-process implementations of the ports for unit testing. decrementInventory is
// atomic within the single-threaded event loop (no await mid-operation), and the
// Locker serialises per key, so the reservation tests exercise the real orchestration.

import { Errors } from '../domain/errors';
import {
  Event,
  Order,
  Payment,
  QueueToken,
  Reservation,
  ReservationStatus,
  Sector,
  User,
} from '../domain/models';
import {
  EventRepository,
  LockHandle,
  Locker,
  QueueRepository,
  RateLimiter,
  ReservationRepository,
  SectorRepository,
  UserRepository,
} from '../usecase/ports';

export class Store {
  users = new Map<string, User>();
  usersByEmail = new Map<string, User>();
  events = new Map<string, Event>();
  sectors = new Map<string, Sector>();
  queue = new Map<string, QueueToken>();
  queueSeq = new Map<string, number>();
  reservations = new Map<string, Reservation>();
  resByIdem = new Map<string, string>();
  orders = new Map<string, Order>();
  orderByIdem = new Map<string, string>();
  orderByRes = new Map<string, string>();
  payments = new Map<string, Payment>();

  putUser(u: User) {
    this.users.set(u.id, u);
    this.usersByEmail.set(u.email, u);
  }
  putEvent(e: Event) {
    this.events.set(e.id, e);
  }
  putSector(s: Sector) {
    this.sectors.set(s.id, s);
  }
}

export class MemUsers implements UserRepository {
  constructor(private s: Store) {}
  async findByEmail(email: string) {
    return this.s.usersByEmail.get(email) ?? null;
  }
  async findById(id: string) {
    return this.s.users.get(id) ?? null;
  }
}

export class MemEvents implements EventRepository {
  constructor(private s: Store) {}
  async list(_cursor: string, limit: number) {
    return { events: [...this.s.events.values()].slice(0, limit), nextCursor: '' };
  }
  async findById(id: string) {
    return this.s.events.get(id) ?? null;
  }
}

export class MemSectors implements SectorRepository {
  constructor(private s: Store) {}
  async listByEvent(eventId: string) {
    return [...this.s.sectors.values()].filter((x) => x.eventId === eventId);
  }
  async findById(id: string) {
    return this.s.sectors.get(id) ?? null;
  }
  async decrementInventory(sectorId: string, qty: number) {
    const x = this.s.sectors.get(sectorId);
    if (!x || x.availableInventory < qty) return false;
    x.availableInventory -= qty;
    return true;
  }
  async incrementInventory(sectorId: string, qty: number) {
    const x = this.s.sectors.get(sectorId);
    if (x) x.availableInventory += qty;
  }
}

const qk = (u: string, e: string) => `${u}|${e}`;

export class MemQueue implements QueueRepository {
  constructor(private s: Store) {}
  async upsert(t: QueueToken) {
    this.s.queue.set(qk(t.userId, t.eventId), { ...t });
  }
  async find(userId: string, eventId: string) {
    return this.s.queue.get(qk(userId, eventId)) ?? null;
  }
  async nextPosition(eventId: string) {
    const p = this.s.queueSeq.get(eventId) ?? 0;
    this.s.queueSeq.set(eventId, p + 1);
    return p;
  }
}

const rk = (u: string, k: string) => `${u}|${k}`;

export class MemReservations implements ReservationRepository {
  constructor(private s: Store) {}
  async create(r: Reservation) {
    const key = rk(r.userId, r.idempotencyKey);
    if (this.s.resByIdem.has(key)) throw Errors.Conflict;
    this.s.reservations.set(r.id, { ...r });
    this.s.resByIdem.set(key, r.id);
  }
  async findById(id: string) {
    return this.s.reservations.get(id) ?? null;
  }
  async findByIdempotencyKey(userId: string, key: string) {
    const id = this.s.resByIdem.get(rk(userId, key));
    return id ? (this.s.reservations.get(id) ?? null) : null;
  }
  async updateStatus(id: string, status: string) {
    const r = this.s.reservations.get(id);
    if (r) r.status = status as ReservationStatus;
  }
  async findExpired(now: Date, limit: number) {
    return [...this.s.reservations.values()]
      .filter((r) => r.status === ReservationStatus.Held && now > r.expiresAt)
      .slice(0, limit);
  }
}

export class MemLocker implements Locker {
  private locks = new Map<string, Promise<void>>();
  async acquire(key: string, _waitMs: number): Promise<LockHandle> {
    // Chain promises per key so holders run one at a time within the event loop.
    const prev = this.locks.get(key) ?? Promise.resolve();
    let release!: () => void;
    const next = new Promise<void>((resolve) => (release = resolve));
    this.locks.set(
      key,
      prev.then(() => next),
    );
    await prev;
    return { release: async () => release() };
  }
}

export class AllowAllRateLimiter implements RateLimiter {
  async allow() {
    return true;
  }
}

File: backends/fastapi/app/adapter/memory.py
"""In-process, async implementations of the ports for unit testing. The decrement is
atomic within the event loop (no await mid-operation), and the locker serialises per
key, so the reservation tests exercise the real use-case orchestration.
"""
from __future__ import annotations

import asyncio
from datetime import datetime, timedelta

from app.domain import errors
from app.domain.models import (
    Event,
    Order,
    Payment,
    QueueToken,
    Reservation,
    Sector,
    User,
)


class Store:
    def __init__(self) -> None:
        self.users: dict[str, User] = {}
        self.users_by_email: dict[str, User] = {}
        self.events: dict[str, Event] = {}
        self.sectors: dict[str, Sector] = {}
        self.queue: dict[tuple[str, str], QueueToken] = {}
        self.queue_seq: dict[str, int] = {}
        self.reservations: dict[str, Reservation] = {}
        self.res_by_idem: dict[tuple[str, str], str] = {}
        self.orders: dict[str, Order] = {}
        self.order_by_idem: dict[tuple[str, str], str] = {}
        self.order_by_res: dict[str, str] = {}
        self.payments: dict[str, Payment] = {}

    def put_user(self, u: User) -> None:
        self.users[u.id] = u
        self.users_by_email[u.email] = u

    def put_event(self, e: Event) -> None:
        self.events[e.id] = e

    def put_sector(self, s: Sector) -> None:
        self.sectors[s.id] = s

    # UserRepository
    async def find_by_email(self, email: str) -> User | None:
        return self.users_by_email.get(email)

    async def find_by_id(self, user_id: str) -> User | None:
        return self.users.get(user_id)

    # EventRepository
    async def list(self, cursor: str, limit: int) -> tuple[list[Event], str]:
        items = list(self.events.values())[:limit]
        return items, ""

    async def event_by_id(self, event_id: str) -> Event | None:
        return self.events.get(event_id)

    # SectorRepository
    async def list_by_event(self, event_id: str) -> list[Sector]:
        return [s for s in self.sectors.values() if s.event_id == event_id]

    async def sector_by_id(self, sector_id: str) -> Sector | None:
        return self.sectors.get(sector_id)

    async def decrement_inventory(self, sector_id: str, qty: int) -> bool:
        s = self.sectors.get(sector_id)
        if s is None or s.available_inventory < qty:
            return False
        s.available_inventory -= qty
        return True

    async def increment_inventory(self, sector_id: str, qty: int) -> None:
        s = self.sectors.get(sector_id)
        if s is not None:
            s.available_inventory += qty

    # QueueRepository
    async def upsert(self, token: QueueToken) -> None:
        self.queue[(token.user_id, token.event_id)] = token

    async def find(self, user_id: str, event_id: str) -> QueueToken | None:
        return self.queue.get((user_id, event_id))

    async def next_position(self, event_id: str) -> int:
        p = self.queue_seq.get(event_id, 0)
        self.queue_seq[event_id] = p + 1
        return p

    # ReservationRepository
    async def create(self, r: Reservation) -> None:
        key = (r.user_id, r.idempotency_key)
        if key in self.res_by_idem:
            raise errors.CONFLICT
        self.reservations[r.id] = r
        self.res_by_idem[key] = r.id

    async def reservation_by_id(self, reservation_id: str) -> Reservation | None:
        return self.reservations.get(reservation_id)

    async def find_by_idempotency_key(self, user_id: str, key: str) -> Reservation | None:
        rid = self.res_by_idem.get((user_id, key))
        return self.reservations.get(rid) if rid else None

    async def update_status(self, reservation_id: str, status: str) -> None:
        r = self.reservations.get(reservation_id)
        if r is not None:
            from app.domain.models import ReservationStatus

            r.status = ReservationStatus(status)

    async def find_expired(self, now: datetime, limit: int) -> list[Reservation]:
        from app.domain.models import ReservationStatus

        out = [
            r
            for r in self.reservations.values()
            if r.status == ReservationStatus.HELD and now > r.expires_at
        ]
        return out[:limit]


# Thin per-aggregate views. The ports reuse method names (find_by_id, create) across
# aggregates, so each gets its own adapter that maps port names onto the store.


class Users:
    def __init__(self, s: Store):
        self.s = s

    async def find_by_email(self, email: str) -> User | None:
        return await self.s.find_by_email(email)

    async def find_by_id(self, user_id: str) -> User | None:
        return await self.s.find_by_id(user_id)


class Events:
    def __init__(self, s: Store):
        self.s = s

    async def list(self, cursor: str, limit: int) -> tuple[list[Event], str]:
        return await self.s.list(cursor, limit)

    async def find_by_id(self, event_id: str) -> Event | None:
        return await self.s.event_by_id(event_id)


class Sectors:
    def __init__(self, s: Store):
        self.s = s

    async def list_by_event(self, event_id: str) -> list[Sector]:
        return await self.s.list_by_event(event_id)

    async def find_by_id(self, sector_id: str) -> Sector | None:
        return await self.s.sector_by_id(sector_id)

    async def decrement_inventory(self, sector_id: str, qty: int) -> bool:
        return await self.s.decrement_inventory(sector_id, qty)

    async def increment_inventory(self, sector_id: str, qty: int) -> None:
        await self.s.increment_inventory(sector_id, qty)


class Queue:
    def __init__(self, s: Store):
        self.s = s

    async def upsert(self, token: QueueToken) -> None:
        await self.s.upsert(token)

    async def find(self, user_id: str, event_id: str) -> QueueToken | None:
        return await self.s.find(user_id, event_id)

    async def next_position(self, event_id: str) -> int:
        return await self.s.next_position(event_id)


class Reservations:
    def __init__(self, s: Store):
        self.s = s

    async def create(self, r: Reservation) -> None:
        await self.s.create(r)

    async def find_by_id(self, reservation_id: str) -> Reservation | None:
        return await self.s.reservation_by_id(reservation_id)

    async def find_by_idempotency_key(self, user_id: str, key: str) -> Reservation | None:
        return await self.s.find_by_idempotency_key(user_id, key)

    async def update_status(self, reservation_id: str, status: str) -> None:
        await self.s.update_status(reservation_id, status)

    async def find_expired(self, now: datetime, limit: int) -> list[Reservation]:
        return await self.s.find_expired(now, limit)


class AllowAllRateLimiter:
    async def allow(self, key: str, limit: int, window: timedelta) -> bool:
        return True


class Locker:
    """Per-key asyncio lock. acquire returns a handle whose release() frees it."""

    def __init__(self) -> None:
        self._locks: dict[str, asyncio.Lock] = {}

    async def acquire(self, key: str, wait: timedelta):
        lock = self._locks.setdefault(key, asyncio.Lock())
        try:
            await asyncio.wait_for(lock.acquire(), timeout=wait.total_seconds())
        except asyncio.TimeoutError:
            return None
        return _Handle(lock)


class _Handle:
    def __init__(self, lock: asyncio.Lock) -> None:
        self._lock = lock

    async def release(self) -> None:
        self._lock.release()
